File: 3439_Reschedule_Meetings_For_Maximum_Free_Time_1.py
from typing import List

class Solution:
    def maxFreeTime(self, eventTime: int, k: int, startTime: List[int], endTime: List[int]) -> int:
        # Step 1: Build free gaps before, between, and after meetings
        gaps = []
        gaps.append(startTime[0])  # before first meeting

        for i in range(1, len(startTime)):
            gaps.append(startTime[i] - endTime[i - 1])  # between meetings

        gaps.append(eventTime - endTime[-1])  # after last meeting

        # Step 2: Use sliding window of size (k+1)
        window = sum(gaps[:k + 1])  # initial window
        maxGap = window

        for i in range(k + 1, len(gaps)):
            window += gaps[i] - gaps[i - (k + 1)]
            maxGap = max(maxGap, window)

        return maxGap

File: 3439_Reschedule_Meetings_For_Maximum_Free_Time_1.cpp
class Solution {
    public:
        int maxFreeTime(int eventTime, int k, vector<int>& startTime, vector<int>& endTime) {
            vector<int> gaps;
    
            // Gap before first meeting
            gaps.push_back(startTime[0]);
    
            // Gaps between meetings
            for (int i = 1; i < startTime.size(); ++i) {
                gaps.push_back(startTime[i] - endTime[i - 1]);
            }
    
            // Gap after last meeting
            gaps.push_back(eventTime - endTime.back());
    
            // Sliding window of size k+1
            int window = 0;
            for (int i = 0; i <= k; ++i)
                window += gaps[i];
    
            int maxGap = window;
    
            for (int i = k + 1; i < gaps.size(); ++i) {
                window += gaps[i] - gaps[i - (k + 1)];
                maxGap = max(maxGap, window);
            }
    
            return maxGap;
        }
    };
    